American Option Dynamic Prog.cpp

```diff
@@ -0,0 +1,148 @@
+			if (j == i)
+				transition_probability(i,j) = notick_prob;
+			if (j == (i+1))
+				transition_probability(i,j) = uptick_prob;
+		}			
+	
+	Matrix V_t(2*no_of_divisions+1,1);
+	// value at expiration
+	for (int i = 1; i <= 2*no_of_divisions+1; i++) 
+		V_t(i,1) = max(0.0, (initial_stock_price*pow(up_factor,i-1-no_of_divisions)) - strike_price);
+	
+	// value at intermediate stages
+	Matrix discounted_one_step_forward_value (2*no_of_divisions+1,1);
+	Matrix value_if_option_is_exercised_now (2*no_of_divisions+1,1);
+	for (int i = no_of_divisions; i > 0; i--) {
+		// going backwards from expiration to zero-time
+		discounted_one_step_forward_value = (transition_probability * V_t)/R;
+		for (int j = 1; j <= 2*no_of_divisions+1; j++)
+			value_if_option_is_exercised_now(j,1) = max(0, (initial_stock_price*pow(up_factor, j-1-no_of_divisions) - strike_price));
+		for (int j = 1; j <= 2*no_of_divisions+1; j++)
+			V_t(j,1) = max(value_if_option_is_exercised_now(j,1), discounted_one_step_forward_value(j,1));
+	}
+	
+	// At the end of this backward movement, the middle-entry is the value of the american option
+	
+	return (V_t(no_of_divisions+1,1));
+}
+
+float american_put_option_dyn_prog()
+{
+	// create the probability matrix
+	Matrix transition_probability(2*no_of_divisions+1, 2*no_of_divisions+1);
+	
+	for (int i = 1; i <= 2*no_of_divisions+1; i++)
+		for (int j = 1; j <= 2*no_of_divisions+1; j++)
+			transition_probability(i,j) = 0.0;
+	
+	// boundary values of the probabilities need to be entered
+	transition_probability(1,1) = 1.0 - uptick_prob;
+	transition_probability(1,2) = uptick_prob;
+	transition_probability(2*no_of_divisions+1,2*no_of_divisions+1) = 1.0 - downtick_prob;
+	transition_probability(2*no_of_divisions+1,2*no_of_divisions) = downtick_prob;
+	
+	for (int i = 2; i <= 2*no_of_divisions; i++)
+		for (int j = 1; j <= 2*no_of_divisions+1; j++) {
+			if (j == (i-1))
+				transition_probability(i,j) = downtick_prob;
+			if (j == i)
+				transition_probability(i,j) = notick_prob;
+			if (j == (i+1))
+				transition_probability(i,j) = uptick_prob;
+		}			
+	
+	Matrix V_t(2*no_of_divisions+1,1);
+	// value at expiration
+	for (int i = 1; i <= 2*no_of_divisions+1; i++) 
+		V_t(i,1) = max(0.0, strike_price - (initial_stock_price*pow(up_factor,i-1-no_of_divisions)));
+	// value at intermediate stages
+	Matrix discounted_one_step_forward_value (2*no_of_divisions+1,1);
+	Matrix value_if_option_is_exercised_now (2*no_of_divisions+1,1);
+	for (int i = no_of_divisions; i > 0; i--) {
+		// going backwards from expiration to zero-time
+		discounted_one_step_forward_value = (transition_probability * V_t)/R;
+		for (int j = 1; j <= 2*no_of_divisions+1; j++)
+			value_if_option_is_exercised_now(j,1) = max(0, strike_price - (initial_stock_price*pow(up_factor, j-1-no_of_divisions)));
+		for (int j = 1; j <= 2*no_of_divisions+1; j++)
+			V_t(j,1) = max(value_if_option_is_exercised_now(j,1), discounted_one_step_forward_value(j,1));
```

Comment:
no need for value_if_option_is_exercised_now, it could be
V_t(j,1) = max(max(0, strike_price - (initial_stock_price*pow(up_factor, j-1-no_of_divisions))), discounted_one_step_forward_value(j,1));